File: src/serialize.js
const escape = (c, special) => {
  if (!special)
    special = ' !"#$%&\'()*+,-./0123456789:;<=>?@[\\]^_`{|}~';
  return (special.indexOf(c) >= 0 ? '\\' : '') + c;
}
const escapeAttr = (c) => escape (c, '\\}');

const topStateChar = (t) => {
  if (typeof(t) === 'string')
    return t;
  switch (t.op) {
  case '+':
  case '-':
  case '*':
    return '(' + makeStateChar(t) + ')';
  default:
    return makeStateChar(t);
  }
}

const makeStateChar = (t) => {
  if (typeof(t) === 'string')
    return t;
  switch (t.op) {
  case 'char':
    return escape(t.char);
  case 'wild':
    return '?';
  case 'any':
    return '*';
  case 'class':
    return '[' + t.chars.map(makeStateChar).join('') + ']';
  case 'negated':
    return '[^' + t.chars.map(makeStateChar).join('') + ']';
  case 'neighborhood':
    return '@' + t.neighborhood + '(' + vecExpr(t.origin) + ')';
  case 'clock':
  case 'anti':
      return '@' + t.op + '(' + vecExpr(t.arg) + ')';
  case 'add':
  case 'sub':
      return '@' + t.op + '(' + vecExpr(t.left) + ',' + vecExpr(t.right) + ')';
  case '+':
  case '-':
  case '*':
  case 'location':
  case 'absdir':
  case 'reldir':
  case 'integer':
  case 'vector':
  case 'state':
  case 'tail':
      return vecExpr(t);
  default:
    throw new Error ("Unrecognized op '" + t.op + "' in " + JSON.stringify(t));
  }
}

const vecExpr = (t) => {
  switch (t.op) {
  case '+':
  case '-':
    return vecExpr(t.left) + t.op + vecExpr(t.right);
  case '*':
    return vecExpr(t.left) + t.op + multiplicativeVecExpr(t.right);
  case 'location':
    return '@' + t.group;
  case 'absdir':
  case 'reldir':
      return '@' + t.dir;
  case 'integer':
    return '@int(' + t.n + ')';
  case 'vector':
    return '@vec(' + t.x + ',' + t.y + ')';
  case 'state':
    return '$' + (t.group || '') + '#' + t.char;
  case 'tail':
    return '$' + (t.group || '') + '#*';
  case 'matrix':
    return '%' + t.matrix;
  default:
    throw new Error ("Unrecognized op '" + t.op + "' in " + JSON.stringify(t));
  }
}
const multiplicativeVecExpr = (t) => {
  const s = vecExpr(t);
  return (t.op === '+' || t.op === '-') ? ('(' + s + ')') : s;
}

const stateSuffix = (t) => {
  return t.state ? ("/" + t.state.map(topStateChar).join('')) : '';
}

const termWithState = (t) => {
  return (typeof(t.type) === 'number' ? 't' : '') + t.type + stateSuffix(t);
}

const lhsTerm = (t) => {
  switch (t.op) {
    case 'any':
      return '*';
    case 'negterm':
      return '^' + lhsTerm(t.term);
    case 'alt':
      return '(' + t.alt.map(lhsTerm).join('|') + ')';
    default:
      return termWithState(t);
    }
}

const addrExpr = (t) => {
  switch (t.op) {
    case 'absdir':
    case 'reldir':
        return '>' + t.dir.toUpperCase() + '>';
    case 'cell':
      return '>' + vecExpr(t.arg) + '>';
    case 'neighbor':
      return '>+' + vecExpr(t.arg) + '>';
    default:
      throw new Error ("Unrecognized op '" + t.op + "' in " + JSON.stringify(t));
      return undefined;
    }
}

const makeLhs = (lhs) => {
  return lhs.slice(1).reduce ((list, curr) => {
    return list + ' ' + (curr.addr ? (addrExpr(curr.addr) + ' ') : '') + lhsTerm(curr);
  }, [lhsTerm(lhs[0])]);
}

const rhsTerm = (t) => {
  if (t.op === 'group')
    return '$' + t.group;
  if (t.op === 'prefix')
    return '$' + t.group + '/' + stateSuffix(t);
  return termWithState(t);
};

const makeRhs = (rhs, sep) => {
  return rhs.map(rhsTerm).join(sep);
};

const fixedPoint6 = (rate) => {
  let s = Math.floor (rate / 1000000), f = Math.floor (rate % 1000000);
  if (f) {
    f = f.toString();
    while (f.length < 6)
      f = '0' + f;
    s = (s + '.' + f).replace(/0+$/g,'');
  }
  return s;
}

const serialize = (rules) => {
  const attrs = (rule) =>
    (rule.rate ? (' rate=' + fixedPoint6(rule.rate)) : '')
    + (rule.sync ? (' sync=' + fixedPoint6(rule.sync)) : '')
    + (rule.command ? (' command={' + escapeAttr(rule.command) + '}') : '')
    + (rule.key ? (' key={' + escapeAttr(rule.key) + '}') : '')
    + (rule.reward ? (' reward=' + rule.reward) : '')
    + (rule.sound ? (' sound={' + escapeAttr(rule.sound) + '}') : '')
    + (rule.caption ? (' caption={' + escapeAttr(rule.caption) + '}') : '');
  return rules.map ((rule) => {
    switch (rule.type) {
      case 'transform': {
        const a = attrs(rule);
        return makeLhs(rule.lhs) + ' : ' + makeRhs(rule.rhs,' ')
        + (a.length ? ("," + a) : "")
        + ".\n";
      }
      case 'inherit':
        return rule.child + ' = ' + rule.parents.join(', ') + ".\n";
      case 'comment':
        return '//' + rule.comment + "\n";
      default:
        throw new Error ("Unrecognized rule type '" + rule.type + "' in " + JSON.stringify(rule));
      }
  }).join("");
}

export { serialize, lhsTerm }

File: python/serialize.py
def escape(c, special=None):
    if not special:
        special = ' !"#$%&\'()*+,-./0123456789:;<=>?@[\\]^_`{|}~'
    return ('\\' if c in special else '') + c

def escape_attr(c):
    return escape(c, '\\}')

def top_state_char(t):
    if isinstance(t, str):
        return t
    if t['op'] in ['+', '-', '*']:
        return '(' + make_state_char(t) + ')'
    return make_state_char(t)

def make_state_char(t):
    if isinstance(t, str):
        return t
    match t['op']:
        case 'char':
            return escape(t['char'])
        case 'wild':
            return '?'
        case 'any':
            return '*'
        case 'class':
            return '[' + ''.join(map(make_state_char, t['chars'])) + ']'
        case 'negated':
            return '[^' + ''.join(map(make_state_char, t['chars'])) + ']'
        case 'neighborhood':
            return '@' + t['neighborhood'] + '(' + vec_expr(t['origin']) + ')'
        case 'clock' | 'anti':
            return '@' + t['op'] + '(' + vec_expr(t['arg']) + ')'
        case 'add' | 'sub':
            return '@' + t['op'] + '(' + vec_expr(t['left']) + ',' + vec_expr(t['right']) + ')'
        case _:
            return vec_expr(t)

def vec_expr(t):
    match t['op']:
        case '+' | '-':
            return vec_expr(t['left']) + t['op'] + vec_expr(t['right'])
        case '*':
            return vec_expr(t['left']) + t['op'] + multiplicative_vec_expr(t['right'])
        case 'location':
            return '@' + t['group']
        case 'absdir' | 'reldir':
            return '@' + t['dir']
        case 'integer':
            return '@int(' + str(t['n']) + ')'
        case 'vector':
            return '@vec(' + str(t['x']) + ',' + str(t['y']) + ')'
        case 'state':
            return '$' + (t['group'] or '') + '#' + t['char']
        case 'tail':
            return '$' + (t['group'] or '') + '#*'
        case 'matrix':
            return '%' + t['matrix']
        case _:
            raise ValueError(f"Unrecognized op '{t['op']}' in {t}")

def multiplicative_vec_expr(t):
    s = vec_expr(t)
    return '(' + s + ')' if t['op'] in ['+', '-'] else s

def state_suffix(t):
    return "/" + ''.join(map(top_state_char, t['state'])) if t.get('state') else ''

def term_with_state(t):
    return ('t' if isinstance(t['type'], int) else '') + str(t['type']) + state_suffix(t)

def lhs_term(t):
    match t['op']:
        case 'any':
            return '*'
        case 'negterm':
            return '^' + lhs_term(t['term'])
        case 'alt':
            return '(' + '|'.join(map(lhs_term, t['alt'])) + ')'
        case _:
            return term_with_state(t)

def addr_expr(t):
    match t['op']:
        case 'absdir' | 'reldir':
            return '>' + t['dir'].upper() + '>'
        case 'cell':
            return '>' + vec_expr(t['arg']) + '>'
        case 'neighbor':
            return '>+' + vec_expr(t['arg']) + '>'
        case _:
            raise ValueError(f"Unrecognized op '{t['op']}' in {t}")

def make_lhs(lhs):
    return ' '.join([lhs_term(lhs[0])] + [(addr_expr(curr['addr']) + ' ' if curr.get('addr') else '') + lhs_term(curr) for curr in lhs[1:]])

def rhs_term(t):
    if t['op'] == 'group':
        return '$' + t['group']
    if t['op'] == 'prefix':
        return '$' + t['group'] + '/' + state_suffix(t)
    return term_with_state(t)

def make_rhs(rhs, sep):
    return sep.join(map(rhs_term, rhs))

def fixed_point6(rate):
    s, f = divmod(rate, 1000000)
    f_str = f"{f:06d}".rstrip('0') or '0'
    return f"{s}.{f_str}" if f else str(s)

def serialize(rules):
    def attrs(rule):
        return ''.join([
            f" rate={fixed_point6(rule['rate'])}" if 'rate' in rule else '',
            f" sync={fixed_point6(rule['sync'])}" if 'sync' in rule else '',
            f" command={{{escape_attr(rule['command'])}}}" if 'command' in rule else '',
            f" key={{{escape_attr(rule['key'])}}}" if 'key' in rule else '',
            f" reward={rule['reward']}" if 'reward' in rule else '',
            f" sound={{{escape_attr(rule['sound'])}}}" if 'sound' in rule else '',
            f" caption={{{escape_attr(rule['caption'])}}}" if 'caption' in rule else '',
        ])
    return ''.join([
        f"{make_lhs(rule['lhs'])} : {make_rhs(rule['rhs'], ' ')},{attrs(rule)}.\n" if rule['type'] == 'transform' else
        f"{rule['child']} = {', '.join(rule['parents'])}.\n" if rule['type'] == 'inherit' else
        f"//{rule['comment']}\n" if rule['type'] == 'comment' else
        raise ValueError(f"Unrecognized rule type '{rule['type']}' in {rule}")
        for rule in rules
    ])


